docs/guides/features/ipsec-offload.md

```diff
@@ -0,0 +1,138 @@
+to prepare system with generated TDI.json and context.json file references.
+
+The /usr/share/stratum/es2k/es2k_skip_p4.conf file must include the fixed
+function configuration reference.
+
+```json
+"fixed_functions" : [
+  {
+   "name": "crypto",
+   "tdi": "/tmp/fixed/tdi.json",
+   "ctx": "/tmp/fixed/crypto-mgr-ctx.json"
+  }
+],
+```
+
+With this configuration updated, you can start infrap4d.
+
+### Compile P4 program
+
+Compile the ipsec-offload program according to the instructions in
+[Compiling P4 programs](/guides/es2k/compiling-p4-programs.md)
+to generate crypto P4 artifacts for programming the pipelines.
+
+### Configure strongSwan
+
+Follow the instructions in the [strongSwan documentation](https://docs.strongswan.org/docs/5.9/index.html)
+to configure host for the use-case selected. This includes details such as
+certificate location, IPs, lifetime thresholds, etc.
+
+### Start IPsec
+
+With the strongSwan application configured, starting IPsec
+(see [ipsec-recipe](https://github.com/ipdk-io/ipsec-recipe) for details) will 
+initiate the pipeline, program the SPD rules as per the P4 program, and
+configure/re-configure SAD entries based on negotiated encryption parameters
+between local and peer system.
+
+## OpenConfig and gNMI messages
+
+This section provides detailed information on OpenConfig model and gNMI
+messages with the expected format. The strongSwan plugin has the following
+details encoded. This is only provided for clarity and information purposes.
+
+### Config SAD message
+
+The P4 Control Plane provides SET and DELETE API calls service to the 
+IPsec-offload [Config SAD message](https://github.com/ipdk-io/openconfig-public/blob/master/release/models/ipsec/openconfig-ipsec-offload.yang#L39-L185)
+at `/ipsec-offload/sad/sad-entry[offload-id=x][direction=y]/config`.
+
+Note that individual leaf nodes of the YANG tree cannot be set and the gRPC
+server expects the full config message as one single message. This design
+was chosen in order to avoid caching of the encryption keys in the
+software stack.
+
+The expected format of the Config SAD message is:
+
+```text
+offload_id: 1,
+direction: true,
+req_id: 2,
+spi: 599,
+ext_seq_num: false,
+anti_replay_window_size: 64,
+protocol_parameters: IPSEC_PROTOCOL_PARAMS_ESP,
+mode: IPSEC_MODE_TRANSPORT,
+esp_payload {
+  encryption {
+    encryption_algorithm: 12,
+    key: "33:65:6b:12:d7:9f:56:63:0b:59:a3:9f:8b:03:d2:c1:b8:51:98:e8",
+    key_len: 20,
+  }
+},
+sa_hard_lifetime {
+  bytes: 200000
+},
+sa_soft_lifetime {
+  bytes: 100000
+}
+```
+
+### Fetch SPI message
+
+The GET gNMI call is used to retrieve the
+[SPI value](https://github.com/ipdk-io/openconfig-public/blob/master/release/models/ipsec/openconfig-ipsec-offload.yang#L292)
+at `/ipsec-offload/ipsec-spi/rx-spi`.
+
+### Key Expiry Notification message
+
+The [gRPC Notification message](https://github.com/ipdk-io/openconfig-public/blob/master/release/models/ipsec/openconfig-ipsec-offload.yang#L308)
```

Comment:
Can we talk about when re-keying is required? Like when a timer expires, etc?

Reply:
This is a well-understood/well-document IPsec (IKE protocol) mechanism. Dont think we need to describe reasonings behind a public protocol in this documentation